crates/ironrdp-connector/src/credssp_sequence.rs

```diff
@@ -0,0 +1,214 @@
+use ironrdp_pdu::write_buf::WriteBuf;
+use ironrdp_pdu::{nego, PduHint};
+use sspi::credssp::{self, ClientState, CredSspClient};
+use sspi::generator::{Generator, NetworkRequest};
+use sspi::negotiate::ProtocolConfig;
+use sspi::KerberosConfig;
+
+use crate::{
+    ClientConnector, ClientConnectorState, ConnectorError, ConnectorErrorKind, ConnectorResult, ServerName, Written,
+};
+
+#[derive(Clone, Copy, Debug)]
+struct CredsspTsRequestHint;
+
+const CREDSSP_TS_REQUEST_HINT: CredsspTsRequestHint = CredsspTsRequestHint;
+
+impl PduHint for CredsspTsRequestHint {
+    fn find_size(&self, bytes: &[u8]) -> ironrdp_pdu::PduResult<Option<usize>> {
+        match sspi::credssp::TsRequest::read_length(bytes) {
+            Ok(length) => Ok(Some(length)),
+            Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => Ok(None),
+            Err(e) => Err(ironrdp_pdu::custom_err!("CredsspTsRequestHint", e)),
+        }
+    }
+}
+#[derive(Clone, Copy, Debug)]
+struct CredsspEarlyUserAuthResultHint;
+
+const CREDSSP_EARLY_USER_AUTH_RESULT_HINT: CredsspEarlyUserAuthResultHint = CredsspEarlyUserAuthResultHint;
+
+impl PduHint for CredsspEarlyUserAuthResultHint {
+    fn find_size(&self, _: &[u8]) -> ironrdp_pdu::PduResult<Option<usize>> {
+        Ok(Some(sspi::credssp::EARLY_USER_AUTH_RESULT_PDU_SIZE))
+    }
+}
+
+pub type CredSspProcessGenerator<'a> = Generator<'a, NetworkRequest, sspi::Result<Vec<u8>>, sspi::Result<ClientState>>;
+
+#[derive(Debug)]
+pub struct CredSspSequence {
+    client: CredSspClient,
+    next_request: Option<credssp::TsRequest>,
+    state: CredSSPState,
+    selected_protocol: nego::SecurityProtocol,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum CredSSPState {
+    CredsspInitial,
+    CredsspReplyNeeded,
+    CredsspEarlyUserAuthResult,
+    Finishied,
+}
+
+impl CredSspSequence {
+    pub fn next_pdu_hint(&self) -> Option<&dyn PduHint> {
+        match self.state {
+            CredSSPState::CredsspInitial => None,
+            CredSSPState::CredsspReplyNeeded => Some(&CREDSSP_TS_REQUEST_HINT),
+            CredSSPState::CredsspEarlyUserAuthResult => Some(&CREDSSP_EARLY_USER_AUTH_RESULT_HINT),
+            CredSSPState::Finishied => panic!(),
+        }
+    }
+
+    pub fn new(
+        connector: &ClientConnector,
+        server_name: impl Into<ServerName>,
```

Comment:
Take a `ServerName` directly, to avoid unnecessary monomorphization.
```suggestion
        server_name: ServerName,
```
See: https://github.com/Devolutions/IronRDP/blob/23777d6b8b9da20f575b861cdc36430a9e1c975a/ARCHITECTURE.md#core-tier